modules/drivers/mongo/src/metabase/driver/mongo/util.clj

```diff
@@ -1,262 +1,93 @@
-    (with-mongo-connection [^com.mongodb.DB conn @(:db (sel :one Table ...))]
-      ...)
-
-    ;; You can use a string instead of a Database
-    (with-mongo-connection [^com.mongodb.DB conn \"mongodb://127.0.0.1:27017/test\"]
-       ...)
-
-  `database-or-connection-string` can also optionally be the connection details map on its own."
-  [[database-binding database] & body]
-  `(let [f# (fn [~database-binding]
-              ~@body)]
-     (if *mongo-connection*
-       (f# *mongo-connection*)
-       (do-with-mongo-connection f# ~database))))
+(defn mongo-client
+  "Create `MongoClient` from `MongoClientSettings`."
+  ^MongoClient [^MongoClientSettings settings]
+  (MongoClients/create settings))
+
+(defn close
+  "Close `client`."
+  [^MongoClient client]
+  (.close client))
+
+(defn database
+  "Get database by its name from `client`."
+  ^MongoDatabase [^MongoClient client db-name]
+  (.getDatabase client db-name))
+
+;; https://mongodb.github.io/mongo-java-driver/4.11/apidocs/mongodb-driver-sync/com/mongodb/client/MongoDatabase.html#runCommand(org.bson.conversions.Bson)
+;; Returns Document
+(defn run-command
+  "Run mongo command."
+  ([^MongoDatabase db cmd & {:as opts}]
+   (let [cmd-doc (mongo.conversion/to-document cmd)]
+     (-> (.runCommand db cmd-doc)
+         (mongo.conversion/from-document opts)))))
+
+;; https://mongodb.github.io/mongo-java-driver/4.11/apidocs/mongodb-driver-sync/com/mongodb/client/MongoClient.html#listDatabaseNames()
+;; returns MongoIterable<String>
+(defn list-database-names
+  "Return vector of names of databases for `client`."
+  [^MongoClient client]
+  (vec (.listDatabaseNames client)))
+
+;; https://mongodb.github.io/mongo-java-driver/4.11/apidocs/mongodb-driver-sync/com/mongodb/client/MongoDatabase.html#listCollectionNames()
+(defn list-collection-names
+  "Return vector of collection names for `db`"
+  [^MongoDatabase db]
+  (vec (.listCollectionNames db)))
+
+(defn collection
+  "Return `MongoCollection` for `db` by its name."
+  ^MongoCollection [^MongoDatabase db coll-name]
+  (.getCollection db coll-name))
+
+;; https://mongodb.github.io/mongo-java-driver/4.11/apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#listIndexes()
+(defn list-indexes
+  "Return vector of Documets describing indexes."
+  [^MongoCollection coll & {:as opts}]
+  (mongo.conversion/from-document (.listIndexes coll) opts))
+
+;; https://mongodb.github.io/mongo-java-driver/4.11/apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#find()
+(defn do-find
+  "Perform find on collection."
+  [^MongoCollection coll
+   & {:keys [limit skip batch-size sort-criteria] :as opts}]
+  (->> (cond-> ^FindIterable (.find coll)
+         limit (.limit limit)
+         skip (.skip skip)
+         batch-size (.batchSize (int batch-size))
+         sort-criteria (.sort (mongo.conversion/to-document sort-criteria)))
+       (mapv #(mongo.conversion/from-document % (select-keys opts [:keywordize])))))
```

Comment:
Wouldn't the following work?

```suggestion
       (mapv #(mongo.conversion/from-document % opts))))
```

Reply:
I wanted to separate `from-document` params from those of `do-find` to keep `from-document`s input "clean". That would make sense if there were using more parameters.